File: take_home/pirates/js/gameMaster.js
const Board = require('./board')
const Island = require('./island')
const CargoShip = require('./cargoShip')
//NOTE: Tight coupling of input and output. Could cause extensibility issues.
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
})

class GameMaster {
    constructor() {
        //NOTE: delegates state management to Board Object and is bad design. state should be floated up to Game Master or at the highest level 
        this.board = new Board();
    }

    turn() {
        this.board.display();
        readline.question("Which direction would you like to move? (1: North, 2: East, 3: South, 4: West): ", direction => {
            //NOTE: why are directions numbers? Seems abitrary and confusing.  Should use letters or capture the arrow key in stdin
            if (direction === '1') {
                //NOTE: confusing code organization.  the board is not doing the moving.
                //BUG: should be moveNorth()
                this.board.moveSouth();
            }
            else if (direction === '2') {
                this.board.moveEast();
            }
            else if (direction === '3') {
                this.board.moveSouth();
            }
            else  {
                //bad defaulting here
                this.board.moveWest();
            }

            const collision = this.board.getCollision()
            if (collision) {
                //instance checking here is a bit clunky
                if (collision instanceof Island) {
                    console.log('You hit an island!')
                    return;
                }
                else if (collision instanceof CargoShip) {
                    console.log('You hit a cargo ship!')
                    this.board.pirate.plunder(collision)
                }
            }
            //NOTE: this "event pump" mechanism is not great. 
            //this looks like a recursive call which can leak memory and blow up the stack
            //alternatives here could be a while loop and/or an event based programming flow
            //event based flow might be preferered as this allows better extensibility of the game
            this.turn()
        })
    }
}

module.exports = GameMaster;

File: take_home/pirates/python/game_master.py
from board import Board
import sys

from island import Island


class GameMaster:
    def __init__(self):
        self.board = Board()

    def turn(self):
        self.board.display()

        sys.stdout.write("Which direction would you like to move?: ")
        sys.stdout.write("1: North")
        sys.stdout.write("2: East")
        sys.stdout.write("3: South")
        sys.stdout.write("4: West")

        direction = input()

        if direction == "1":
            self.board.move_south()
        elif direction == "2":
            self.board.move_east()
        elif direction == "3":
            self.board.move_south()
        else:
            self.board.move_west()

        collision = self.board.get_collision()
        if collision is not None:
            if isinstance(collision, Island):
                print("You crashed into land!")
                sys.exit()
            else:
                print("Avast, there be plunder!")
                self.board.pirate.plunder(collision)

        self.turn()
